[Xiangcheng] iP

Xczheng0105 opened the pull request:
### **Trackie**

> “You don’t need a new plan for next year. You need ~a commitment~ TRACKIE BOT RIGHT NOW.” – [Seth Godin(modified)](https://www.chanty.com/blog/productivity-quotes/)

Trackie is a bot to help you track your tasks.
- simple
- effective
- reliable

(And did I forget to mention, it's _**FREE!**_)

All you need to do is:
1. Launch the bot
2. Add you tasks
3. Let it track 'em all 😎 

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```ruby
public static void main(String[] args) {
        new Trackie("src").run();
    }
```
Features:
- [x] Add tasks
- [x] Remove tasks
- [x] Mark/unmark tasks
- [ ] GUI interface (Work in Progress)

<dl>
  <dt>If you can type fast</dt>
  <dd>Then this is the right task tracker for you!</dd>
</dl>

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Xczheng0105 committed:
refactor from "Duke" to trackie, add greeting and exit messages

Xczheng0105 committed:
add support for echo, design improvements

Xczheng0105 committed:
add support for add tasks and list tasks feature, remove the echo feature

Xczheng0105 committed:
add functionalities for marking and unmarking tasks, some refactoring of code.

Xczheng0105 committed:
add support for adding different types of tasks - todo, deadline, event

Xczheng0105 committed:
add support for semi-automated testing.

Xczheng0105 committed:
add support for error handling

Xczheng0105 committed:
add support for deleting tasks, update testcases

Xczheng0105 committed:
Add support for reading and writing to a data file

Xczheng0105 committed:
Re-implement functionality for saving tasks

Xczheng0105 committed:
Fix some bugs related to loading of tasks on startup

Xczheng0105 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Fix some bugs related to loading of tasks on startup
  Re-implement functionality for saving tasks
  Add support for reading and writing to a data file

Xczheng0105 committed:
Change the way of storing deadlines

Xczheng0105 committed:
Merge branch 'branch-level-8'

* branch-level-8:
  Change the way of storing deadlines

Xczheng0105 committed:
Add more OOP

Xczheng0105 committed:
Merge branch 'branch-A-moreOOP'

* branch-A-moreOOP:
  Add more OOP

Xczheng0105 committed:
Sort the classes into packages

Xczheng0105 committed:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Sort the classes into packages

Xczheng0105 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

Professional-Procrastinat0r commented on src/main/java/trackie/commands/AddCommand.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public void execute(TaskList tasklist, Ui ui, Storage storage) {
+        try {
+            switch (arguments[0]) {
```
Should not have indentation for `case` clauses.

Professional-Procrastinat0r commented on src/main/java/trackie/parsing/Parser.java:
```diff
@@ -0,0 +1,31 @@
+                return new ListCommand();
+            case "bye":
+                return new ExitCommand();
+            default:
```
Good use of default to catch invalid commands

Professional-Procrastinat0r commented on src/main/java/trackie/commands/AddCommand.java:
```diff
@@ -0,0 +1,32 @@
+package trackie.commands;
```
Good adherence to package and import statements orderings

Professional-Procrastinat0r commented on src/main/java/trackie/parsing/Parser.java:
```diff
@@ -0,0 +1,31 @@
+public class Parser {
+    public static Command parseCommand(String input) {
+        String[] arguments = input.split(" ");
+        switch (arguments[0]) {
```
Likewise, `case` clauses should not be indented

Professional-Procrastinat0r commented on src/main/java/trackie/ui/Trackie.java:
```diff
@@ -0,0 +1,34 @@
+import trackie.storage.Storage;
+import trackie.storage.TaskList;
+
+public class Trackie {
```
Can consider adding javadoc comments to your classes for clarity

Professional-Procrastinat0r commented on src/main/java/trackie/commands/AddCommand.java:
```diff
@@ -0,0 +1,32 @@
+package trackie.commands;
+
+import trackie.storage.Storage;
```
Good adherence to explicitly import specific classes

Xczheng0105 committed:
Add gradle support

Xczheng0105 committed:
Add support for junit testing and minor fixes

Xczheng0105 committed:
Add support for jar file and fix some minor bugs

maertan commented on src/main/java/trackie/commands/Command.java:
```diff
@@ -0,0 +1,24 @@
+import trackie.ui.Ui;
+
+public abstract class Command {
+    protected String[] arguments;
```
Good naming of arguments as plural due to it being an array.

maertan commented on src/main/java/trackie/commands/Command.java:
```diff
@@ -0,0 +1,24 @@
+        this.isExit = isExit;
+    }
+
+    public boolean isExit() {
```
Good naming of the boolean - starts with "is".

maertan commented on src/main/java/trackie/parsing/Parser.java:
```diff
@@ -0,0 +1,27 @@
+package trackie.parsing;
+
+import trackie.commands.*;
```
Perhaps you should have explicit import statements rather than using *

maertan commented on src/main/java/trackie/storage/Storage.java:
```diff
@@ -0,0 +1,95 @@
+
+public class Storage {
+    private TaskList taskList;
+    private File fPtr;
```
Perhaps `fPtr` could be written out as a full word, `filePointer` I'm guessing? So that it would be easier for other people reading your code to understand.

maertan commented on src/main/java/trackie/storage/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+            throw new TrackieException("Correct usage: event [desc] /from [start] /to [end]");
+        }
+        //retrieve the start time
+        StringBuilder sb2 = new StringBuilder();
```
Perhaps you could use another way of naming to distinguish the StringBuilders rather than using numbers.

maertan commented on src/main/java/trackie/tasks/Task.java:
```diff
@@ -0,0 +1,42 @@
+
+public abstract class Task {
+    protected String description;
+    protected boolean completed;
```
You could consider naming this boolean something like `isCompleted` so that it sounds like a boolean.

maertan reviewed (commented):
Good job on overall code quality! Here are some suggestions in the comments below:

Xczheng0105 committed:
Include JavaDoc for some functions

Xczheng0105 committed:
Tweak the code to adhere to the coding standard

Xczheng0105 committed:
Add support for finding tasks that match a query

Xczheng0105 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add support for finding tasks that match a query

Xczheng0105 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to adhere to the coding standard

Xczheng0105 committed:
Merge branch 'branch-A-JavaDoc'

Xczheng0105 committed:
Fix some minor bugs

Xczheng0105 committed:
Add Iteration 1 - Echoing the user

Xczheng0105 committed:
Add Iteration 2 - Add dialog box for Trackie's response

Xczheng0105 committed:
Refactor Trackie to use FXML

Xczheng0105 committed:
Link the backend to the frontend

Xczheng0105 committed:
Fix minor bugs and delete the Ui class

Xczheng0105 committed:
Delete the Ui class (did not manage to delete properly beforehand)

Xczheng0105 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Delete the Ui class (did not manage to delete properly beforehand)

Xczheng0105 committed:
Add assertions for the add command

Xczheng0105 committed:
Refactor the code

Some methods in the code are over 100 lines long.

This may hinder code readability.

Let's refactor the code and extract some smaller methods from the
long method.

This will make the code more readable.

Xczheng0105 committed:
Change the name of one variable to adhere to coding standard

Xczheng0105 committed:
Merge pull request #2 from Xczheng0105/heads/A-CodeQuality

Refactoring of code

Xczheng0105 committed:
Date and Time : update the way to store start and end times

Start and end times are stored as strings in an event task.

This can lead to nonsensical inputs.

Let's change the type of start and end time to a LocalDateTime.

This way, start and end times will not contain blatantly
nonsensical inputs.

Xczheng0105 committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/trackie/commands/AddCommand.java

Xczheng0105 committed:
Change some keywords for commands

Xczheng0105 committed:
Change the date and time syntax

For event classes, the default date and time syntax to follow is
quite awkward and complicated.

This can result in a bad user experience.

Let's change the date and time syntax to something that is more
user-friendly.

This way, dates and times will be presented in a more user-readable
format.

Xczheng0105 committed:
Clean up commands

Xczheng0105 committed:
Merge branch 'branch-C-FriendlierSyntax'

Xczheng0105 committed:
Improve the UI

Xczheng0105 committed:
Give the bot a more unique personality

Xczheng0105 committed:
Improve testing for Trackie bot

Xczheng0105 committed:
Merge branch 'branch-A-Personality' into branch-A-MoreTesting

# Conflicts:
#	src/main/java/trackie/commands/InvalidCommand.java

Xczheng0105 committed:
Merge branch 'branch-A-MoreTesting'

Xczheng0105 committed:
Update the error message for invalid command

Xczheng0105 committed:
Fix a minor bug

Xczheng0105 committed:
Update the README file

Xczheng0105 committed:
Fix formatting issues

Xczheng0105 committed:
Fix a minor formatting issue

Xczheng0105 committed:
Fix minor formatting issues

Xczheng0105 committed:
Update the representative screenshot

Xczheng0105 committed:
Provide more JavaDocs

Xczheng0105 committed:
Update docs/README.md

Xczheng0105 committed:
Update docs/README.md

Xczheng0105 committed:
Update docs/README.md

Xczheng0105 committed:
Update docs/README.md

Xczheng0105 committed:
Fix some minor bugs
